Add `/itchysats/offer/2.0.0`

luckysori opened the pull request:
Fixes https://github.com/itchysats/itchysats/issues/2459.
Fixes #2670.

The maker supports versions 1.0.0 and 2.0.0. We keep a copy of the old version of the protocol, but remove a lot of the code that is not actually necessary (the taker side and the tests) and adapt what's left so that it is easier to integrate it in `maker` together with `/itchysats/offer/2.0.0`.

Because the maker is the dialer in the offer protocol, it means that currently all their takers will get doubled offers (one per version) and will ignore the ones that correspond to the protocol they don't support. To prevent us from generating false positive warnings we filter out the `yamux::NegotiationError::Failed` error.

Additionally:

- Rework the way the offers are modelled in different components across the entire codebase.
- The `projection::Actor` now takes a `Role` parameter, so that we can do away with the awkward `Origin` type in the new offer protocol.
- Simplify `model::Offer`, which is not actually used anywhere where backwards-compatibility matters.
- Calculate `oracle_event_id` based on `settlement_interval` when constructing `Offer`. As opposed to passing two arguments that could be out of sync, like we used to do.

---

TODOs:

- [x] Update changelog.
- [x] Update commit message to mention that the new version of the offer protocol inherently supports new contract symbols.
- [x] Consider splitting into more patches.
- [x] Check if we have some unused dependencies after these patches.

luckysori committed:
chore(model): Remove Display impl for OpeningFee

It's only used in the tests, and it wasn't printing the units. If we
use the `Display` implementation of `bitcoin::Amount` it reads better.

luckysori committed:
fix: Fix typos in comments

Restioson commented on daemon/src/projection.rs:
```diff
@@ -1158,8 +1208,24 @@ impl Actor {
+            .filter_map(|offer| match CfdOffer::new(offer, self.role) {
+                Ok(offer) => Some(offer),
+                Err(e) => {
+                    tracing::warn!("Failed to map new offer: {e:#}");
```
I think "failed to map" is probably a bit vague

Restioson commented on xtra-libp2p-offer/src/current/maker.rs:
```diff
@@ -0,0 +1,139 @@
+
+                // This code path should not be hit, but in case we run into an error this sleep
+                // prevents a continuous endless loop of restarts.
+                tokio_extras::time::sleep(Duration::from_secs(2)).await;
```
Is this referring to the actor restarts? If so, couldn't this be better served by a supervisor restart policy with timeout?

luckysori replied:
I only moved this code around, but unfortunately the diff isn't very good :(

I don't fully understand why we need this, but what you suggest sounds like a great idea!!

luckysori replied:
I've created https://github.com/itchysats/itchysats/issues/2717 to track this!

da-kami replied:
I think this is actually outdated and should be removed because there is already a more global strategy in place. But I think it is a good idea to tackle it outside of this PR. Thanks for adding the ticket!

da-kami commented on daemon/src/order/maker.rs:
```diff
@@ -54,7 +52,7 @@ pub struct Actor {
     decision_senders: HashMap<OrderId, oneshot::Sender<protocol::Decision>>,
     db: sqlite_db::Connection,
-    latest_offers: MessageChannel<xtra_libp2p_offer::maker::GetLatestOffers, Option<MakerOffers>>,
+    latest_offers: MessageChannel<xtra_libp2p_offer::maker::GetLatestOffers, Vec<model::Offer>>,
```
❓Why do we use a `Vec` here and not a `Hashmap` that indexes per `symbol` as well? Or hide the whole abstraction behind a struct that hides how this is implemented?

luckysori replied:
I wanted to keep the offer actor as simple as possible. I didn't see the need to use a `HashMap` because a simple list works, but we can consider it if there is a good reason for it.

luckysori replied:
Taking as feedback to wrap Vec<model::Offer> around a offer protocol newtype.

da-kami replied:
Yeah, that way we can make the `impl` of `model::Offers` explicit to the combinations that can be set at the moment and the internal representation does not matter.

For the `Offers` sent over the wire we can consider a model that uses a `Hashmap` - or our `model::Offers` can ensure that the `protocol::Offers` is sane when converting into it. We can opt to keep a `Vec` internally for both, but then then the accessors should ensure that one can only set sane combinations and no duplicates.

da-kami commented on daemon/src/projection.rs:
```diff
@@ -828,26 +829,70 @@ impl Tx {
+    fn send_offer_update(&self, offers: &Offers) -> Result<()> {
+        let OffersByContractSymbol { btc_usd, eth_usd } = split_offers_by_contract_symbol(offers);
+
+        self.order.btc_usd.send(btc_usd)?;
```
🙃 `order` should actually be renamed to `offer` - this might be because this is based on an old branch?

luckysori replied:
Nope this is the same in master AFAIK.

da-kami replied:
I think I missed that one when cleaning up 😅 - we can do it in a follow up :)

da-kami commented on daemon/src/projection.rs:
```diff
@@ -828,26 +829,70 @@ impl Tx {
 struct State {
     network: Network,
     quote: Option<xtra_bitmex_price_feed::Quote>,
+    offers: Offers,
```
❓ Keeping the state for the "current / latest offers" up to date in multiple places in the codebase can be error prone. We already have the `latest_offers` in multiple places.  
What is the motivation for this (avoid one more message between actors...?)? 
I do think it makes sense to push it and store it, but I wonder if this could lead to unwanted side-effects where the state is not up to date in some cases.

luckysori replied:
> What is the motivation for this (avoid one more message between actors...?)?

No messages have changed. The `projection::Actor` still hears about the most recent offers via the `Update` message. What has had to change is the fact that we no longer necessarily send _all_ the current offers from maker to taker every time we run the `/itchysats/offer`. It's perfectly acceptable for the maker to only tell us about ETHUSD offers if that's the only thing that has changed.

With that in mind, I thought that it would be useful to keep a holistic view of the offers in the `projection::Actor` and update it every time we receive an `Update` message. Thinking about it now I am not totally sure that it is necessary. I wonder if it's always sufficient to relay the offers that have changed via the feed. What do you think? Keep in mind that this is different to what we used to do. Before we would always receive _all_ the offers (all BTCUSD) and pump that into the feed.

> I do think it makes sense to push it and store it, but I wonder if this could lead to unwanted side-effects where the state is not up to date in some cases.

With the current functionality in this PR what we are doing is:

1. Receive an update with the new offers.
2. Combine our state with the new offers.
3. Pushing all offers to the feeds.

I think maybe we just need:

1. Receive an update with the new offers.
2. Relay the new offers no matter what.

It's probably bad to send events for offers that haven't actually changed! I'm not 100% sure about how we're supposed to use SSE though, so maybe you have more thoughts on this.

da-kami replied:
I'd say it's better to tackle the state of the `Offer` throughout the application (and multiple actors) in a separate follow-up. You can decide if you want to keep it in the projection as you implemented. 
Tbh I don't think it will make a big difference, given that offers are regularly published anyway, resulting in new `OfferId`s which in turn will require an update.

da-kami commented on daemon/src/projection.rs:
```diff
@@ -828,26 +829,70 @@ impl Tx {
+    fn send_offer_update(&self, offers: &Offers) -> Result<()> {
+        let OffersByContractSymbol { btc_usd, eth_usd } = split_offers_by_contract_symbol(offers);
+
+        self.order.btc_usd.send(btc_usd)?;
```
❓ Why do we need two channels for different offers per symbol? With https://github.com/itchysats/itchysats/pull/2679 I think one channel should be enough or am I missing something?

luckysori replied:
The PR you reference hasn't been merged yet. We'll have to adapt this once that happens.

da-kami commented on xtra-libp2p-offer/src/deprecated.rs:
```diff
@@ -0,0 +1,4 @@
+pub mod maker;
+pub mod protocol;
+
+pub const PROTOCOL: &str = "/itchysats/offer/1.0.0";
```
❓ Why is this called `deprecated`, when for `rollover` we added versions, i.e. `v_1_0_0` and `v_2_0_0` - I think it would help the code base to have some consistency here.

luckysori replied:
Because it's much clearer (I introduced the original horrible terminology). I've adapted the rollover one in a separate PR.

luckysori replied:
Correction: I forgot to actually open the PR 5 days ago! [Here it is](https://github.com/itchysats/itchysats/pull/2706).

da-kami reviewed (commented):
I am not sure about some of these changes, I think it would help to do some design work before implementing a refactoring for this protocol. Was this re-design planned or did it evolve while writing the code? 

Happy to talk about it together.

luckysori committed:
logs(endpoint): Do not emit error event if opening a substream fails

Failing to open a substream will almost always result in the caller
handling the error, which will almost certainly be logged. Therefore,
there is no need for emitting a redundant error event via the
`tracing::instrument` macro.

Furthermore, this will allow callers to decide which errors should be
logged and which can be silenced.

luckysori committed:
chore: Remove unused semver dependency

luckysori committed:
feat(network): Add /itchysats/offer/2.0.0

The `/itchysats/offer/2.0.0` protocol offers proper support for
sending and receiving offers with different contract symbols. The
design of the message exchanged between maker and taker has changed to
make this easier. These are the key differences:

- Before, we only allowed the maker to send a pair of offers that were
supposed to reference the same contract symbol, which was sufficient
for BTCUSD. Now, we give freedom to the maker to send a list of
offers. The taker must combine the new offers with the old ones to
derive a full view of the latest offers.

- Before, parts of the message were optional to allow the maker to
"disable" certain offers and even rollovers. This feature is not
needed since both the maker has other mechanisms in place to reject
orders and rollovers. Now, all fields are mandatory (sending an empty
list is technically possible, but it doesn't mean anything for the
taker).

The maker supports versions 1.0.0 and 2.0.0 side by side. We maintain
version 1.0.0 for the maker in a `deprecated` module, but remove a lot
of the code that is no longer necessary (the taker side and the
tests). We have adapted what remains so that it is easier to integrate
it in `maker` together with the new protocol.

Because the maker is the dialer in the offer protocol, it means that
currently all their takers will get doubled offers (one per version)
and will ignore the ones that correspond to the protocol they don't
support. To prevent us from generating false positive warnings on the
maker side, we filter out the `yamux::NegotiationError::Failed` error.

Additionally:

- Rework the way the offers are modelled in different components
across the entire codebase.
- The `projection::Actor` now takes a `Role` parameter, so that we can
do away with the awkward `Origin` type in the new offer protocol.
- Simplify `model::Offer`, which is not actually used anywhere
where backwards-compatibility matters.
- Calculate `oracle_event_id` based on `settlement_interval` when
constructing `Offer`. As opposed to passing two arguments that could
be out of sync, like we used to do.

da-kami commented on daemon/src/projection.rs:
```diff
@@ -1263,46 +1240,12 @@ impl From<xtra_bitmex_price_feed::Quote> for Quote {
-        };
-        MakerOffers { long, short }
-    }
+    pub btcusd_long: Option<CfdOffer>,
```
🤔 This is a bit annoying, but on the `projection` this is acceptable to flatten out instead of using e.g. a `HashMap` as representation.

luckysori replied:
It's not what I wanted to do originally, but it is the simplest thing I could come up with for the time being. Luckily we can change and break this as much as we want, even after releasing.

da-kami commented on xtra-libp2p-offer/src/current/protocol.rs:
```diff
@@ -0,0 +1,178 @@
+}
+
+#[derive(Clone, Serialize, Deserialize, PartialEq, Debug)]
+pub(crate) struct Offers(Vec<Offer>);
```
🤔 After another look I am not sure I agree with this model.

This model *implies* that the maker sends sane data.
A maker could send multiple offers for the same `contract_symbol` and `position` with this API.
Can't we be more explicit here?

This is what popped into my head:

* long and short are fixed
* the contract_symbold should scale, i.e. we might have more in the future

This leads me to a model like this:

```rust
pub struct MakerOffers {
   long: HashMap<ContractSymbol, Offer>,
   short: HashMap<ContractSymbol, Offer>
}
```

Additionally I would add a constructor that makes it explicit that one always has to set them for both symbols.
Alternatively we could also go for explicit setters, but I think the constructor it better to ensure we always handle the `Option`.

And getters for the respective one. 

```rust
impl MakerOffers {
    pub new(
        btcusd_long: Option<Offer>, 
        btcusd_short:Option<Offer>, 
        ethusd_long:Option<Offer>, 
        ethusd_short: Option<Offer>
    ) -> Self {
        if let Some(btcusd_long) = btcusd_long {
            self.long.insert(ContractSymbol::BtcUsd, btcusd_long);
        }

        ...
    }
}
```

This will scale as good over the wire, but is more explicit about the data. We can roll with this simpler model of `Vec` if you think it's not worth it, but I think it would be better to be explicit.

---

It is then questionable of the `Offer` should have a position; but I would be OK with that redundancy. The API to set long/short within `Offers` can ensure that long is always set to `long` internally.

---

If you find the `HashMap` too cumbersome we can still model the internals using a `Vec` but I think we should still restrict what can be set in that `Vec`. At the moment the code allows too arbitrary things, because we map from one `Vec` into another, I think at least the model should ensure sanity by e.g. a constructor that restricts the combinations to what is possible.

luckysori replied:
> 🤔 After another look I am not sure I agree with this model.
> 
> This model _implies_ that the maker sends sane data. A maker could send multiple offers for the same `contract_symbol` and `position` with this API. Can't we be more explicit here?
> 
> This is what popped into my head:
> 
>     * long and short are fixed
> 
>     * the contract_symbold should scale, i.e. we might have more in the future
> 
> 
> This leads me to a model like this:
> 
> ```rust
> pub struct MakerOffers {
>    long: HashMap<ContractSymbol, Offer>,
>    short: HashMap<ContractSymbol, Offer>
> }
> ```

I had something similar to this at some point, but I decided against it because you don't actually get that much for the price of making the message more complicated. The taker still has to make sense of the offers they just received together with the offers they already knew about. It's not any more complicated to iterate through a simple list applying the same logic. In the end the taker will always have to:

1. Combine offers received and offers stored.
2. Only keep one copy per `(ContracSymbol, Position)`, always choosing the newest one.

> It is then questionable of the `Offer` should have a position; but I would be OK with that redundancy. The API to set long/short within `Offers` can ensure that long is always set to `long` internally.

I would only opt for something like what you suggest if the `Offer` type sent over the wire omits both the position and the contract symbol. Otherwise the taker would still have to check that offers are correctly categorised on reception. And we would have to maybe ignore offers that are incorrectly categorised?

> Additionally I would add a constructor that makes it explicit that one always has to set them for both symbols. 

I don't see the need for this. Particularly not at the protocol level. Why is it important to limit what offers can be sent when? I expect the maker to want to send offers for only one contract symbol most of the time.

> If you find the `HashMap` too cumbersome we can still model the internals using a `Vec` but I think we should still restrict what can be set in that `Vec`. At the moment the code allows too arbitrary things, because we map from one `Vec` into another, I think at least the model should ensure sanity by e.g. a constructor that restricts the combinations to what is possible.

I am not fully seeing your concern here. Whether the maker sends 4 new offers for all the supported combinations or, for example, 20 copies of the same offer, the taker will still have to process them in the same way: keep the newest per `(ContractSymbol, Position)` pair.

da-kami replied:
> I am not fully seeing your concern here. Whether the maker sends 4 new offers for all the supported combinations or, for example, 20 copies of the same offer, the taker will still have to process them in the same way: keep the newest per (ContractSymbol, Position) pair.

As long as the `maker_creation_timestamp` is sane and allow sorting we will be able to pick a "newest", otherwise it will get tricky. The taker will have an error-path to deal with that is: I received an offer with different values for the same symbol + position + timestamp - which one shall the taker pick?
 
But fair enough, this is somewhat fabricated.  I think you might be right that the cost of modelling it more complicated it not worth it. This model will scale and that's what's important.

luckysori commented on daemon-tests/tests/main/offer.rs:
```diff
@@ -35,10 +35,10 @@ async fn taker_receives_offer_from_maker_on_publication() {
-    assert_eq_offer(published.btcusd_long.clone(), received.btcusd_long.clone());
-    assert_eq_offer(published.btcusd_long.clone(), received.btcusd_long.clone());
     assert_eq_offer(published.btcusd_long, received.btcusd_long);
+    assert_eq_offer(published.btcusd_short, received.btcusd_short);
```
Haha thanks!

luckysori commented on xtra-libp2p-offer/src/current/maker.rs:
```diff
@@ -0,0 +1,139 @@
+
+#[xtra_productivity]
+impl Actor {
+    async fn handle_connection_established(
+        &mut self,
+        msg: endpoint::ConnectionEstablished,
+        ctx: &mut xtra::Context<Self>,
+    ) {
+        tracing::trace!("Adding newly established connection: {:?}", msg.peer_id);
+        self.connected_peers.insert(msg.peer_id);
+        self.send_offers(msg.peer_id, ctx).await;
+    }
```
I've noticed that this is not exactly what we want. Here the maker is only sharing with the taker the very last offers sent over the wire.

I think we want to be able to tell them about all the relevant offers that they have missed. To do so we would need to introduce more complex state in the maker actor i.e. a `HashMap<(ContractSymbol, Position), Offer>`. I spent some time trying to work on this, which also motivated me to model the wire message similarly (as you suggested above, @da-kami).

It gets quite complicated and I don't want to spend more time on this until we have ETHUSD almost working. To move forward I would just create a ticket and add it to the milestone. When we get to tackle it we can decide if we want to do some more remodelling depending on how much time we have.

And I've stashed my changes locally in case we want to revisit them specifically.

da-kami replied:
As long as *all* the latest know offers (i.e. all symbol + position combinations supported) are being sent I think that is sufficient. Currently `takers` don't really care about anything else.

da-kami commented on daemon/src/projection.rs:
```diff
@@ -789,15 +786,20 @@ impl Tx {
-    fn send_offer_update(&self, offer: MakerOffers) {
-        self.offer.send(offer).unwrap_or_default();
+    fn send_offer_update(&self, offers: MakerOffers) -> Result<()> {
+        tracing::info!(?offers, "Sending maker offers to feed");
```
❓Did you add this log here because it is the only point in the application where `maker` and `taker` behave the same when handling "new offers"?

luckysori replied:
I think I was debugging and this slipped through. I will make sure I remove any such log statements.

da-kami replied:
I think it's actually good to introduce a log statement that logs the offers - I don't think we have one since the new protocol was introduced. So feel free to keep one somewhere :)

da-kami commented on daemon/src/projection.rs:
```diff
@@ -1112,8 +1142,24 @@ impl Actor {
+            })
+            .collect_vec();
+
+        self.state.update_offers(new_offers);
```
Just to re-iterate: I don't think updating the `offers` as part of the state here is particularly useful. Feel free to just assemble the `MakerOffers` locally here and then send out an update. My understanding is that we always send out a complete update anyway, so I am not sure keeping `offers` as part of the state has any advantage.

luckysori replied:
If we don't compose the state here we will only send the latest offers (these will reference a single currency pair at a time at the moment) to the feed.

I wanted to keep the pre-existing feed behaviour: on every update we send everything we know, even if it's redundant. It's not efficient (and I think we should not do this eventually), but I didn't want to introduce any unexpected changes.

da-kami replied:
Right, because the maker does not send the offers alltogeher, but sends them by symbol - even thought the new protocol would allow sending them all together (but it's questionable if that is something we would want).
Ok, then let's keep the state for now and tackle this in follow up(s).

da-kami commented on maker/src/routes.rs:
```diff
@@ -60,8 +60,10 @@ pub async fn maker_feed(
         let offers = rx_offers.borrow().clone();
-        yield Event::json(&offers.long).event("long_offer");
-        yield Event::json(&offers.short).event("short_offer");
+        yield Event::json(&offers.btcusd_long).event("btcusd_long_offer");
```
🤔 Previously I did not notice that you opted for splitting this on the `event` level now. That's what I had suggested originally, but we discarded it because it was somewhat not scalable. 
I still think it is the cleanest solution when it comes to having less decision in the UI so I am not against this change. Still wondering what motivated you to change this :)

luckysori replied:
It's the only way that I could keep the UI behaviour without introducing bigger changes. Eventually I think we want something different, but this is good enough now imo.

da-kami reviewed (approved):
Thanks for all the great work. I'll unblock this :)
